src/System.Private.CoreLib/src/System/Text/Utf8Span.Conversion.cs

```diff
@@ -0,0 +1,321 @@
+        {
+            // TODO_UTF8STRING: Reduce allocations in this code path.
+
+            ReadOnlySpan<char> normalized = this.ToString().Normalize(normalizationForm);
+            OperationStatus status = Utf8.FromUtf16(normalized, destination, out int _, out int bytesWritten, replaceInvalidSequences: false, isFinalBlock: true);
+
+            Debug.Assert(status == OperationStatus.Done || status == OperationStatus.DestinationTooSmall, "Normalize shouldn't have produced malformed Unicode string.");
+
+            if (status != OperationStatus.Done)
+            {
+                bytesWritten = -1; // "destination too small"
+            }
+
+            return bytesWritten;
+        }
+
+        /// <summary>
+        /// Converts this <see cref="Utf8Span"/> to a <see langword="char[]"/>.
+        /// </summary>
+        public unsafe char[] ToCharArray()
+        {
+            if (IsEmpty)
+            {
+                return Array.Empty<char>();
+            }
+
+            // TODO_UTF8STRING: Since we know the underlying data is immutable, well-formed UTF-8,
+            // we can perform transcoding using an optimized code path that skips all safety checks.
+            // We should also consider skipping the two-pass if possible.
+
+            fixed (byte* pbUtf8 = &DangerousGetMutableReference())
+            {
+                byte* pbUtf8Invalid = Utf8Utility.GetPointerToFirstInvalidByte(pbUtf8, this.Length, out int utf16CodeUnitCountAdjustment, out _);
+                Debug.Assert(pbUtf8Invalid == pbUtf8 + this.Length, "Invalid UTF-8 data seen in buffer.");
+
+                char[] asUtf16 = new char[this.Length + utf16CodeUnitCountAdjustment];
+                fixed (byte* pbUtf16 = &asUtf16.GetRawSzArrayData())
+                {
+                    OperationStatus status = Utf8Utility.TranscodeToUtf16(pbUtf8, this.Length, (char*)pbUtf16, asUtf16.Length, out byte* pbUtf8End, out char* pchUtf16End);
+                    Debug.Assert(status == OperationStatus.Done, "The buffer changed out from under us unexpectedly?");
+                    Debug.Assert(pbUtf8End == pbUtf8 + this.Length, "The buffer changed out from under us unexpectedly?");
+                    Debug.Assert(pchUtf16End == ((char*)pbUtf16) + asUtf16.Length, "The buffer changed out from under us unexpectedly?");
+
+                    return asUtf16;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Converts this <see cref="Utf8Span"/> instance to its UTF-16 equivalent, writing the result into
+        /// the buffer <paramref name="destination"/>.
+        /// </summary>
+        /// <returns>
+        /// The number of bytes written to <paramref name="destination"/>, or -1 if <paramref name="destination"/>
+        /// is not large enough to hold the result of the transcoding operation.
+        /// </returns>
+        public int ToChars(Span<char> destination)
```

Comment:
>ToChars [](start = 19, length = 7)

nit: The other API called ToCharArray, wouldn't this make sense to call it something like FillCharSpan?